### PR TITLE
Added official json schema test suite for uniqueitems

### PR DESCRIPTION
**Pull request by sorenbronsted**
@TimothyBlynJacobs I have added the official json schema test suite for uniqueitems

Trac ticket: https://core.trac.wordpress.org/ticket/48821

### CODE REVIEW

**Comment by TimothyBJacobs** [comment]
Thanks @sorenbronsted!

Support for min and max items was added in https://core.trac.wordpress.org/changeset/47923.

Let's also skip `additionalItems` and positional `items` support please.

**Comment by sorenbronsted** [comment]
> Let's also skip `additionalItems` and positional `items` support please.

Can I ask why you want to skip theses things, when I have put the time and effort into making all the tests work?

**Comment by TimothyBJacobs** [comment]
I think its a complicated feature to implement and is fairly unintuitive for users to use. I think an object with named properties is almost always a better DUX.

In the latest spec its also been removed from `items` and moved to a clearer `tupleItems`.

File: src/wp-includes/rest-api.php [hunk 1]
```diff
@@ -1305,6 +1305,24 @@ function rest_validate_value_from_schema( $value, $args, $param = '' ) {
 			/* translators: 1: Parameter, 2: Number. */
 			return new WP_Error( 'rest_invalid_param', sprintf( __( '%1$s must contain at most %2$s items.' ), $param, number_format_i18n( $args['maxItems'] ) ) );
 		}
+
+		if ( isset( $args['uniqueItems'] ) && $args['uniqueItems'] ) {
+			foreach ( $value as $index => &$object ) {
+				if ( is_array( $object ) ) {
+					ksort( $object );
```
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api.php:1312]: Does this work for deeply nested objects?

I think we should extract a `rest_is_array_unique` function.

**Comment by TimothyBJacobs** [comment]
Something like this:

```php
/**
 * Checks if an array is made up of unique items.
 *
 * @since 5.5.0
 *
 * @param array $array The array to check.
 * @return bool True if the array contains unique items, false otherwise.
 */
function rest_validate_array_contains_unique_items( $array ) {
	$seen = array();

	foreach ( $array as $item ) {
		$stabilized = rest_stabilize_value( $item );
		$key        = maybe_serialize( $stabilized );

		if ( ! isset( $seen[ $key ] ) ) {
			$seen[ $key ] = 1;

			continue;
		}

		return false;
	}

	return true;
}

/**
 * Stabilizes a value following JSON Schema semantics.
 *
 * For lists, order is preserved. For objects, properties are reordered alphabetically.
 *
 * @since 5.5.0
 *
 * @param mixed $value The value to stabilize. Must already be sanitized. Objects should have been converted to arrays.
 * @return mixed The stabilized value.
 */
function rest_stabilize_value( $value ) {
	if ( is_scalar( $value ) || is_null( $value ) ) {
		return $value;
	}

	if ( is_object( $value ) ) {
		_doing_it_wrong( __FUNCTION__, __( 'Cannot stabilize objects. Convert the object to an array first.' ), '5.5.0' );

		return $value;
	}

	ksort( $value );

	foreach ( $value as $k => $v ) {
		$value[ $k ] = rest_stabilize_value( $v );
	}

	return $value;
}
```

**Comment by TimothyBJacobs** [comment]
A possible patch:

```patch
diff --git a/src/wp-includes/rest-api.php b/src/wp-includes/rest-api.php
index b2ea6e9654..8154d03af3 100644
--- a/src/wp-includes/rest-api.php
+++ b/src/wp-includes/rest-api.php
@@ -54,7 +54,7 @@ function register_rest_route( $namespace, $route, $args = array(), $override = f
 		_doing_it_wrong(
 			'register_rest_route',
 			sprintf(
-				/* translators: %s: rest_api_init */
+			/* translators: %s: rest_api_init */
 				__( 'REST API routes must be registered on the %s action.' ),
 				'<code>rest_api_init</code>'
 			),
@@ -1438,6 +1438,63 @@ function rest_handle_multi_type_schema( $value, $args, $param = '' ) {
 	return $best_type;
 }
 
+/**
+ * Checks if an array is made up of unique items.
+ *
+ * @since 5.5.0
+ *
+ * @param array $array The array to check.
+ * @return bool True if the array contains unique items, false otherwise.
+ */
+function rest_validate_array_contains_unique_items( $array ) {
+	$seen = array();
+
+	foreach ( $array as $item ) {
+		$stabilized = rest_stabilize_value( $item );
+		$key        = serialize( $stabilized );
+
+		if ( ! isset( $seen[ $key ] ) ) {
+			$seen[ $key ] = 1;
+
+			continue;
+		}
+
+		return false;
+	}
+
+	return true;
+}
+
+/**
+ * Stabilizes a value following JSON Schema semantics.
+ *
+ * For lists, order is preserved. For objects, properties are reordered alphabetically.
+ *
+ * @since 5.5.0
+ *
+ * @param mixed $value The value to stabilize. Must already be sanitized. Objects should have been converted to arrays.
+ * @return mixed The stabilized value.
+ */
+function rest_stabilize_value( $value ) {
+	if ( is_scalar( $value ) || is_null( $value ) ) {
+		return $value;
+	}
+
+	if ( is_object( $value ) ) {
+		_doing_it_wrong( __FUNCTION__, __( 'Cannot stabilize objects. Convert the object to an array first.' ), '5.5.0' );
+
+		return $value;
+	}
+
+	ksort( $value );
+
+	foreach ( $value as $k => $v ) {
+		$value[ $k ] = rest_stabilize_value( $v );
+	}
+
+	return $value;
+}
+
 /**
  * Validate a value based on a schema.
  *
@@ -1492,10 +1549,12 @@ function rest_validate_value_from_schema( $value, $args, $param = '' ) {
 
 		$value = rest_sanitize_array( $value );
 
-		foreach ( $value as $index => $v ) {
-			$is_valid = rest_validate_value_from_schema( $v, $args['items'], $param . '[' . $index . ']' );
-			if ( is_wp_error( $is_valid ) ) {
-				return $is_valid;
+		if ( isset( $args['items'] ) ) {
+			foreach ( $value as $index => $v ) {
+				$is_valid = rest_validate_value_from_schema( $v, $args['items'], $param . '[' . $index . ']' );
+				if ( is_wp_error( $is_valid ) ) {
+					return $is_valid;
+				}
 			}
 		}
 
@@ -1509,22 +1568,9 @@ function rest_validate_value_from_schema( $value, $args, $param = '' ) {
 			return new WP_Error( 'rest_invalid_param', sprintf( __( '%1$s must contain at most %2$s items.' ), $param, number_format_i18n( $args['maxItems'] ) ) );
 		}
 
-		if ( isset( $args['uniqueItems'] ) && $args['uniqueItems'] ) {
-			foreach ( $value as $index => &$object ) {
-				if ( is_array( $object ) ) {
-					ksort( $object );
-				}
-			}
-			$unique = array_map(
-				function ( $e ) {
-					return var_export( $e, true );
-				},
-				$value
-			);
-			if ( count( $value ) !== count( array_unique( $unique ) ) ) {
-				/* translators: 1: Parameter */
-				return new WP_Error( 'rest_invalid_param', sprintf( __( '%1$s has duplicate items.' ), $param ) );
-			}
+		if ( ! empty( $args['uniqueItems'] ) && ! rest_validate_array_contains_unique_items( $value ) ) {
+			/* translators: 1: Parameter */
+			return new WP_Error( 'rest_invalid_param', sprintf( __( '%1$s has duplicate items.' ), $param ) );
 		}
 	}
 
@@ -1615,7 +1661,7 @@ function rest_validate_value_from_schema( $value, $args, $param = '' ) {
 			return new WP_Error(
 				'rest_invalid_param',
 				sprintf(
-					/* translators: 1: Parameter, 2: Number of characters. */
+				/* translators: 1: Parameter, 2: Number of characters. */
 					_n( '%1$s must be at least %2$s character long.', '%1$s must be at least %2$s characters long.', $args['minLength'] ),
 					$param,
 					number_format_i18n( $args['minLength'] )
@@ -1627,7 +1673,7 @@ function rest_validate_value_from_schema( $value, $args, $param = '' ) {
 			return new WP_Error(
 				'rest_invalid_param',
 				sprintf(
-					/* translators: 1: Parameter, 2: Number of characters. */
+				/* translators: 1: Parameter, 2: Number of characters. */
 					_n( '%1$s must be at most %2$s character long.', '%1$s must be at most %2$s characters long.', $args['maxLength'] ),
 					$param,
 					number_format_i18n( $args['maxLength'] )
@@ -1647,7 +1693,7 @@ function rest_validate_value_from_schema( $value, $args, $param = '' ) {
 	// The "format" keyword should only be applied to strings. However, for backward compatibility,
 	// we allow the "format" keyword if the type keyword was not specified, or was set to an invalid value.
 	if ( isset( $args['format'] )
-		&& ( ! isset( $args['type'] ) || 'string' === $args['type'] || ! in_array( $args['type'], $allowed_types, true ) )
+		 && ( ! isset( $args['type'] ) || 'string' === $args['type'] || ! in_array( $args['type'], $allowed_types, true ) )
 	) {
 		switch ( $args['format'] ) {
 			case 'hex-color':
@@ -1815,7 +1861,7 @@ function rest_sanitize_value_from_schema( $value, $args, $param = '' ) {
 
 	// This behavior matches rest_validate_value_from_schema().
 	if ( isset( $args['format'] )
-		&& ( ! isset( $args['type'] ) || 'string' === $args['type'] || ! in_array( $args['type'], $allowed_types, true ) )
+		 && ( ! isset( $args['type'] ) || 'string' === $args['type'] || ! in_array( $args['type'], $allowed_types, true ) )
 	) {
 		switch ( $args['format'] ) {
 			case 'hex-color':
diff --git a/tests/phpunit/tests/rest-api/rest-schema-validation.php b/tests/phpunit/tests/rest-api/rest-schema-validation.php
index fae968b40e..37c03be6d1 100644
--- a/tests/phpunit/tests/rest-api/rest-schema-validation.php
+++ b/tests/phpunit/tests/rest-api/rest-schema-validation.php
@@ -906,11 +906,9 @@ class WP_Test_REST_Schema_Validation extends WP_UnitTestCase {
 		$this->assertWPError( rest_validate_value_from_schema( 'foobar', $schema ) );
 	}
 
-	/**
-	 * Data providor for uniqueitems tests
-	 * @return Generator
-	 */
-	public function data_uniqueitems() {
+	public function data_unique_items() {
+		$all_types = array( 'object', 'array', 'null', 'number', 'integer', 'boolean', 'string' );
+
 		// the following test suites is not supported at the moment
 		$skip   = array(
 			'uniqueItems with an array of items',
@@ -919,6 +917,9 @@ class WP_Test_REST_Schema_Validation extends WP_UnitTestCase {
 			'uniqueItems=false with an array of items and additionalItems=false',
 		);
 		$suites = json_decode( file_get_contents( __DIR__ . '/json_schema_test_suite/uniqueitems.json' ), true );
+
+		$tests = array();
+
 		foreach ( $suites as $suite ) {
 			if ( in_array( $suite['description'], $skip, true ) ) {
 				continue;
@@ -929,26 +930,36 @@ class WP_Test_REST_Schema_Validation extends WP_UnitTestCase {
 			}
 			// items is required for our implementation
 			if ( ! isset( $suite['schema']['items'] ) ) {
-				$suite['schema']['items'] = array( 'type' => '' );
+				$suite['schema']['items'] = array(
+					'type'  => $all_types,
+					'items' => array(
+						'type' => $all_types,
+					),
+				);
 			}
 			foreach ( $suite['tests'] as $test ) {
-				yield array( $test, $suite );
+				$tests[] = array( $test, $suite );
 			}
 		}
+
+		return $tests;
 	}
 
 	/**
 	 * @ticket 48821
 	 *
-	 * @dataProvider data_uniqueitems
+	 * @dataProvider data_unique_items
 	 */
-	public function test_uniqueitems( $test, $suite ) {
+	public function test_unique_items( $test, $suite ) {
 		$test_description = $suite['description'] . ': ' . $test['description'];
 		$message          = $test_description . ': ' . var_export( $test['data'], true );
+
+		$valid = rest_validate_value_from_schema( $test['data'], $suite['schema'] );
+
 		if ( $test['valid'] ) {
-			$this->assertTrue( rest_validate_value_from_schema( $test['data'], $suite['schema'] ), $message );
+			$this->assertTrue( $valid, $message );
 		} else {
-			$this->assertWPError( rest_validate_value_from_schema( $test['data'], $suite['schema'] ), $message );
+			$this->assertWPError( $valid, $message );
 		}
 	}
 }
```

**Comment by TimothyBJacobs** [comment]
Merged.

**Pull request closed without merging**